### DISCUSSION TITLE
contributing/development/compiling/compiling_for_web

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#148] [User-contributed notes]
# contributing/development/compiling/compiling_for_web

Requirements: To compile export templates for the Web, the following is required: Emscripten 3.1.39+., Python 3.6+., SCons 3.1.2+ build system.. Building export templates: Before starting, confirm ...

https://docs.godotengine.org/en/stable/contributing/development/compiling/compiling_for_web.html

**Comment by Brennanzuz** [comment]
Typo under [GDExtension](https://docs.godotengine.org/en/stable/contributing/development/compiling/compiling_for_web.html#gdextension):

`Finally, rename the zip archives to web_dlink_release.zip and web_dlink_release.zip` should be `web_dlink_release.zip and web_dlink_debug.zip`.

**Comment by Swarkin** [comment]
For me, scons only showed web as a compilation option once I ran an admin command prompt, terminal didn't work.

**Reply by Calinou** [reply]
Which operating system are you on? There's no requirement to use administrator to compile for any platform. On top of being insecure, this is discouraged as it'll cause file ownership issues on Linux/macOS when the build process creates files.

**Reply by Swarkin** [reply]
I'm using Windows, tried running `emsdk.bat activate latest`followed by `emsdk_env.bat`in the Terminal app which didn't show up in the scons target list. I then attempted it in a cmd admin prompt and scons did show it. I haven't tested whether admin is necessary, I can do so later today.

**Reply by Swarkin** [reply]
Admin doesnt seem to be required, but a regular cmd is. Powershell does not work for me.

**Reply by Calinou** [reply]
I guess what happens is that a `.bat` (cmd) script can't set environment variables correctly within a PowerShell session. I'm still a bit surprised about this, I recall this working correctly in other apps.

**Reply by jacobcoughenour** [reply]
I was able to get it working on Windows after running `emsdk.bat activate latest --permanent` which writes to the PATH environment variable for the system and not just the current session.

**Comment by FBalerio** [comment]
I've been attempting to export a Web template on Windows 11, and had been getting incosnsistent error messages regarding "worker.js". In the console, it looks like this:
```
scons: building terminated because of errors.
[100%] Install file: "bin\godot.web.template_release.wasm32.wrapped.js" as "bin\.web_zip\godot.js"
[100%] Install file: "bin\godot.web.template_release.wasm32.worker.js" as "bin\.web_zip\godot.worker.js"
scons: *** [bin\.web_zip\godot.worker.js] The system cannot find the file specified
scons: building terminated because of errors.
```

I finally got it to work by switching from emscripten's latest version to `3.1.64`. If you run into this error, try that version  instead of the latest one or the one recommended by this article, which also produced errors.

**Reply by ghost** [reply]
Had the same issue on Linux with installed latest (currently 3.1.70). Installing 3.1.64 helped

**Comment by ghost** [comment]
I have installed `emscripten` via `emsdk` on Linux. And I usually run `scons` through `sudo` so it didn't show `web` as supported platform. If you encounter the same problem make sure your settings are available for the root if you do the same

**Comment by Meorge** [comment]
I'm trying to run a custom build of the web editor on macOS in Chrome. I noticed that the docs say

> access `godot.tools.html` and you should be able to test the web editor this way

I have no `godot.tools.html`, but I do have `godot.editor.html`. Opening it, it looks like the web editor is working, but some errors get printed to the JavaScript console. Once I try to open the actual editor from this window, I get an error `TypeError: Cannot read properties of undefined (reading 'msg_send_queue')` and the editor itself never loads.

```
Uncaught ReferenceError: ___GODOT_ENSURE_CROSSORIGIN_ISOLATION_HEADERS___ is not defined
    at service.worker.js:14:46

Uncaught (in promise) TypeError: Failed to register a ServiceWorker for scope ('http://127.0.0.1:8060/godot.web.editor.dev.wasm32/') with script ('http://127.0.0.1:8060/godot.web.editor.dev.wasm32/service.worker.js'): ServiceWorker script evaluation failed

Godot Engine v4.4.dev.custom_build.7b39fcc23 (2024-11-28 04:45:00 UTC) - https://godotengine.org

OpenGL API OpenGL ES 3.0 (WebGL 2.0 (OpenGL ES 3.0 Chromium)) - Compatibility - Using Device: WebKit - WebKit WebGL

godot.editor.js:5034 Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'msg_send_queue')
    at Object.sendmsg (godot.editor.js:5034:14)
    at Object.write (godot.editor.js:4640:28)
    at Object.write (godot.editor.js:3478:42)
    at ___syscall_sendto (godot.editor.js:5428:17)
    at godot.web.editor.dev.wasm32.wasm.sendto (godot.web.editor.dev…-7ecfb11e:0x37fb2ca)
    at godot.web.editor.dev.wasm32.wasm.send (godot.web.editor.dev…-7ecfb11e:0x37fb2b7)
    at godot.web.editor.dev.wasm32.wasm.__netlink_enumerate (godot.web.editor.dev…-7ecfb11e:0x37ccb3f)
    at godot.web.editor.dev.wasm32.wasm.__rtnetlink_enumerate (godot.web.editor.dev…-7ecfb11e:0x37ccad2)
    at godot.web.editor.dev.wasm32.wasm.getifaddrs (godot.web.editor.dev…-7ecfb11e:0x37ca1a2)
    at godot.web.editor.dev.wasm32.wasm.IPUnix::get_local_interfaces(HashMap<String, IP::…asm32.wasm-7ecfb11e)
```

While the changes I've made shouldn't have caused any huge issues, I'll try building from master and seeing if that has the same issue, and if so I'll file a bug report for it.

**Reply by Calinou** [reply]
> I have no `godot.tools.html`, but I do have `godot.editor.html`.

Pull request opened:

- https://github.com/godotengine/godot-docs/pull/10334

**Comment by pewcworrell** [comment]
I also had issues when using dlink_enabled=yes. Installing 3.1.64 helped with some. After that, I was getting an error for Unknown option '--enable-bulk-memory-opt'. I added the argument for production=yes to my template, and that did the trick..

**Reply by Meorge** [reply]
I ran into the same issue compiling the editor, and adding `production=yes` allowed it to get *further* than before, but now it seems to have hung on a cache-related step (either that or it's just taking a very long time)...

**Comment by enetheru** [comment]
Tried to compile 4.3  on windows 11 using emscripten version 3.1.39 as listed in 4.3 docs and it fails with 
`Errors with TypeError: MOZ_TO_ME[node.type] is not a function`
changed emscripten version to 3.1.64 to match GitHub CI for 4.3 branch, and it compiles successfully.
If the CI is 3.1.64, can the documentation be updated to match?

**Comment by etherealxx** [comment]
⚠️**For Git Bash user**
if during installation/activation of enscripten it says python not installed, first type this command:
```
export EMSDK_PYTHON="$(which python)"
```
and then re run the activate and source commands.
Either re run that export command everytime you want to build, or put that command on your bash profile file (i think it's on `"C:\Program Files\Git\etc\profile.d\bash_profile.sh"`)

**Comment by FeralBytes** [comment]
Now requires Emscripten 4.0.11 per this PR https://github.com/godotengine/godot/pull/107460
